src/main/java/fi/helsinki/cs/tmc/core/communication/TmcServerCommunicationTaskFactory.java

```diff
@@ -78,7 +81,33 @@ public void setSettings(TmcSettings settings) {
         this.settings = settings;
     }
 
-    private URI getCourseListUrl() {
+    /**
+     * Returns a Callable that calls the given Callable.
+     *
+     * <p>If the call fails, the auth is refreshed and the call is done again.</p>
+     *
+     * @param <T> return type of the callable
+     * @param callable Callable to be wrapped
+     * @return Callable that wraps the given callable
+     */
+    private <T> Callable<T> tryCallCallable(final Callable<T> callable) {
+        if (!settings.getUsingOauth()) {
+            return callable;
+        }
+        return new Callable<T>() {
+            @Override
+            public T call() throws Exception {
+                try {
+                    return callable.call();
+                } catch (Exception e) {
+                    oauth.refreshToken();
```

Comment:
At minimum we need to refresh it on `FailedHttpResponseException` when status code is 403 or 401.
But I am unsure what all others we would need to prepare for. There could be more status codes or exceptions. The ones mentioned are caused by authorize! and normal no access (forbidden) which I feel could be enough. Cant think of additional cases where token needs refreshing as those are the ones that get triggered when trying to access stuff as a guest or wrong user with no access.